pkg/catalog/catalog.go

```diff
@@ -2768,6 +2744,53 @@ func (c *Catalog) listRepositoriesHelper(ctx context.Context) ([]*graveler.Repos
 	return repos, nil
 }
 
+func (c *Catalog) VerifyLinkAddress(repository, branch, path, physicalAddress string) error {
+	address, signature, found := strings.Cut(physicalAddress, LinkAddressSigningDelimiter)
+	if !found {
+		return fmt.Errorf("address is not signed: %w", graveler.ErrLinkAddressInvalid)
+	}
+
+	stringToVerify := getAddressToSign(repository, branch, path, address)
+	decodedSig, err := base64.URLEncoding.DecodeString(signature)
+	if err != nil {
+		return fmt.Errorf("malformed address signature: %s: %w", stringToVerify, graveler.ErrLinkAddressInvalid)
+	}
+
+	h := hmac.New(sha256.New, []byte(c.signingKey))
+	h.Write([]byte(stringToVerify))
+	calculated := h.Sum(nil)
+	if !hmac.Equal(calculated, decodedSig) {
+		return fmt.Errorf("invalid address signature: %w", block.ErrInvalidAddress)
+	}
+	creationTime, err := c.PathProvider.ResolvePathTime(address)
+	if err != nil {
+		return err
+	}
+
+	if time.Since(creationTime) > LinkAddressTime {
+		return graveler.ErrLinkAddressExpired
+	}
+	return nil
+}
+
+func (c *Catalog) signAddress(logicalAddress string) string {
+	// create a new HMAC by defining the hash type and the key
+	h := hmac.New(sha256.New, []byte(c.signingKey))
+	// compute the HMAC
+	h.Write([]byte(logicalAddress))
+	dataHmac := h.Sum(nil)
+	return base64.URLEncoding.EncodeToString(dataHmac) // Using url encoding to avoid "/"
+}
+
+func getAddressToSign(repository, branch, path, physicalAddress string) string {
+	return repository + branch + path + physicalAddress
```

Comment:
Yes

Reply:
Not sure.  Why not just sign the actual generated physical address?

Reply:
We said that signature should be dependant on repo, branch and path (i.e. you can only use a linked address with a single logical path)

Reply:
> Yes

@N-o-Z how so? It's just string concatenation..

Reply:
OK, to answer both questions:
1. If we sign only the physical address, then users can user generated linked addresses of repo1/branch2/path3 in repo4/branch5/path6
2. Because physical address is always unique, the prefix for the address to sign will always be unique and therefore we will never get to a situation of:
`getAddressToSign("aa", "bb", "cc", "dd")`
`getAddressToSign("a", "ab", "bc", "cdd")`

However, regarding 2, if it makes more sense I can add some separator between the parts of the signature